### PR TITLE
[Merged by Bors] - chore(Algebra/Module): split off `NoZeroSMulDivisors`

### PR DESCRIPTION
**Pull request by Vierkantor**
The definition of `NoZeroSMulDivisors` doesn't require us to know about modules. By splitting off this definition, we can streamline the downstream import graph a bit.

In the process, generalize a few instances from `Module` to `SMulWithZero`.

New files:
 * `NoZeroSMulDivisors/Defs.lean`: definition and basic results that don't require `Module`
 * `NoZeroSMulDivisors/Basic.lean`: basic results that require `Module`
 * `NoZeroSMulDivisors/Pi.lean`: results on `NoZeroSMulDivisors _ (_ -> _)` that don't require `Module`
 * `NoZeroSMulDivisors/Prod.lean`: results on `NoZeroSMulDivisors _ (_ x _)` that don't require `Module`

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by Vierkantor** [comment]
Hmm, the downstream changes are not as impressive as I was hoping. Still, a few files with drastically reduced imports.

**Review by eric-wieser — COMMENTED** [review]
It's not immediately clear to me whether the `Pi` and `Prod` file are a good idea; putting every instance of `Foo Bar` in `Foo/Bar.lean` is a recipe for users never being able to find the instance they need! Then again, we already do this for Prod and Pi elsewhere...

**Comment by jcommelin** [comment]
My gut feeling is that this mild nuisance is worth the trade off. The two PRs that depend on this (#17913 and #17919) give a quite a nice reduction in import counts.

**Comment by kim-em** [comment]
bors merge

File: Mathlib/Algebra/NoZeroSMulDivisors/Defs.lean [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anne Baanen, Yury Kudryashov, Joseph Myers, Heather Macbeth, Kim Morrison, Yaël Dillies
+-/
+import Mathlib.Algebra.SMulWithZero
+
+/-!
+# `NoZeroSMulDivisors`
+
+This file defines the `NoZeroSMulDivisors` class, and includes some tests
+for the vanishing of elements (especially in modules over division rings).
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/NoZeroSMulDivisors/Defs.lean:12]: I don't think this is right anymore

**Pull request closed without merging**